concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -414,12 +431,12 @@ checkTimeoutCertificateSignature tsmGenesis toKeys TimeoutCertificate{..} =
     Bls.verifyAggregateHybrid msgsKeys (theTimeoutSignature tcAggregateSignature)
   where
     msgsKeys =
-        [ (timeoutSignatureMessageBytes TimeoutSignatureMessage{tsmRound = tcRound, ..}, toKeys fs)
-          | (tsmQCRound, fs) <- finalizerRoundsList tcFinalizerQCRounds
+        [ (timeoutSignatureMessageBytes TimeoutSignatureMessage{tsmRound = tcRound, tsmEpoch = tcEpoch, ..}, toKeys fs)
+          | ((tsmQCRound, tsmQCEpoch), fs) <- finalizerRoundsList tcFinalizerQCRounds
```

Comment:
The `toKeys` function needs to change. (By the way, is it implemented somewhere ? I couldn't find the source by searching through the files for that word. But maybe it is in submodule that I haven't loaded.) It should take the epoch as input as well, so that it knows which set of keys to retrieve. That was the reason for including the epoch of the best QC in the `TimeoutSignatureMessage`. I believe it shouldn't be possible for there to be more than two different epochs in the `tcFinalizerQCRounds`, and under certain conditions there should only be one epoch and we can forget the old keys. But we still need to work those details out.

Reply:
It's just a function being passed into the `checkTimeoutCertificateSignature`. With this signature `(FinalizerSet  -> [BakerAggregationVerifyKey])`. I have spiced it up a bit so now it is declared as 
`(FinalizerSet -> Epoch -> [BakerAggregationVerifyKey]).

Reply:
I think exactly how this function retrieves the keys may be handled differently in future when we have a better understanding of how we retrieve the finalization committee(s) involved.

Reply:
Won't spend more time on this now. I put the epoch in the function there for now.